### PR TITLE
feat: add typescript introspector

### PR DESCRIPTION
**Pull request by TomChv**
Add a program to introspect typescript files and convert it to a GraphQL schema.
Enable decorators in Node SDK.
Add analysis function.
Add typescript compiler.
Add decorators to declare expose object and methods to a GQL schema.
Add unit tests

TODO:
- [ ] Add more tests
- [ ] Convert analysed metadata to GraphQL schema

Related to DEV-3029

### CODE REVIEW

**Comment by helderco** [comment]
> declare expose object and methods to a GQL schema

There's no need to build a GraphQL schema. Extending the API is done via API calls.

File: sdk/nodejs/introspector/analysis/analysis.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+ * Analyse the list of Typescript File using the Typescript compiler API.
+ *
+ * This function introspect files and returns metadata of their class and
+ * functions that should be exposed to the Dagger API.
+ *
+ * WARNING(28/11/23): This does NOT include arrow style function.
+ *
+ * @param files List of Typescript files to introspect.
+ */
+export function analysis(files: string[]): Metadata {
```
Comment (helderco) [inline sdk/nodejs/introspector/analysis/analysis.ts:23]: Why is it needed to load source files for introspection? Python relies on the SDK importing the user's module from a conventionalized location (currently `main`), which will execute the decorators just by importing. The decorators then add the decorated object to a registry which is available during execution time.

From the user's main file, other files can be imported there.

File: sdk/nodejs/introspector/decorators/decorators.ts
Comment (helderco) [inline sdk/nodejs/introspector/decorators/decorators.ts:1]: These decorators are just placeholders? I'd expect to see them adding the decorated object to a "registry" and process that later on.

File: sdk/nodejs/introspector/test/analysis.spec.ts [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+            {
+              name: "bar",
+              doc: "Return Bar object",
+              returnType: "Bar",
+              params: [],
+            },
+          ],
+        },
+      ],
+      functions: [],
```
Comment (helderco) [inline sdk/nodejs/introspector/test/analysis.spec.ts:99]: What's this, top-level functions? I see no test/example for it.

File: sdk/nodejs/introspector/test/analysis.spec.ts [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    }
+
+    assert.deepEqual(result, expected)
+  })
+
+  it("should analyse classes' properties to keep a state", async function () {
```
Comment (helderco) [inline sdk/nodejs/introspector/test/analysis.spec.ts:151]: Not important but "analyse" is British English, while "analyze" is US spelling.

File: sdk/nodejs/introspector/test/testdata/state/state.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    public packages: string[]
+
+    ctr: Container
+
+    /**
+     * Returns a base Alpine container
+     * @param version version to use (default to: 3.16.2)
+     */
+    @fct
+    base(version?: string): Alpine {
```
Comment (helderco) [inline sdk/nodejs/introspector/test/testdata/state/state.ts:24]: Why not put the default in the signature? 
```suggestion
    base(version: string = "3.16.2"): Alpine {
```
Then you'll be able to register the arg's [`defaultValue`](https://docs.dagger.io/current/sdk/nodejs/reference/classes/api_client_gen.FunctionArg#defaultvalue), which will show in the CLI automatically.

If that value needs to be shared in other places it can be made a constant.

File: sdk/nodejs/introspector/test/analysis.spec.ts [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+              name: "add",
+              returnType: "number",
+              doc: "",
+              params: [
+                {
+                  name: "a",
+                  typeName: "number",
+                  doc: "",
+                  optional: true,
+                },
```
Comment (helderco) [inline sdk/nodejs/introspector/test/analysis.spec.ts:268]: Correctly identified as optional, but could add the defaultValue here.

File: sdk/nodejs/tsconfig.json [hunk 6]
```diff
@@ -12,6 +12,7 @@
     "module": "ES2022",
     "skipLibCheck": true,
     "forceConsistentCasingInFileNames": true,
+    "experimentalDecorators": true,
```
Comment (helderco) [inline sdk/nodejs/tsconfig.json:15]: TypeScript 5.0 was released in March with [stable support for decorators](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#decorators). The experimental decorators is from a much older and less capable proposal.

**Review by helderco — COMMENTED** [review]
How about object fields? They're implicit while object and function are explicit. The ability to easily make it private is ok but you'll actually need 3 scopes:
- exposed in the api, serialized in result and inputs
- not exposed in the api but still serializable
- private only to language code, not seen in serialization and not exposed to api

The reason Python has `dagger.mod.field` is because it not only helps with these three scopes but also for allowing an alternative name for the field in the API in case it's needed.

File: sdk/nodejs/introspector/test/analysis.spec.ts (hunk 5 above)
Comment (TomChv) [reply sdk/nodejs/introspector/test/analysis.spec.ts:268]: Why? It is not supported by GraphQL asap?

File: sdk/nodejs/introspector/test/analysis.spec.ts (hunk 2 above)
Comment (TomChv) [reply sdk/nodejs/introspector/test/analysis.spec.ts:99]: It's for top level function, it will not be supported by the first version but the field exist for a later implementation

File: sdk/nodejs/introspector/decorators/decorators.ts
Comment (TomChv) [reply sdk/nodejs/introspector/decorators/decorators.ts:1]: I thought about it once I checked on your implementation, I need to check if it's possible to introspect from the decorators itself, I'm not sure about that

File: sdk/nodejs/introspector/decorators/decorators.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+export function fct(
```
[the remark this answers is not in the document]
Comment (TomChv) [reply sdk/nodejs/introspector/decorators/decorators.ts:1]: Ok I can change that!

File: sdk/nodejs/tsconfig.json (hunk 6 above)
Comment (TomChv) [reply sdk/nodejs/tsconfig.json:15]: True, I need to update the typescript version though, I'll check that and see if I can fix it

**Comment by TomChv** [comment]
> How about object fields? They're implicit while object and function are explicit. 

Make sense, I didn't knew if it was important or not but I can add a field

File: sdk/nodejs/tsconfig.json (hunk 6 above)
Comment (TomChv) [reply sdk/nodejs/tsconfig.json]: Resolved in [27477bc](https://github.com/dagger/dagger/pull/6159/commits/27477bc051006373f25a0ad00364ef1e6bc95bea)

File: sdk/nodejs/introspector/test/analysis.spec.ts (hunk 5 above)
Comment (TomChv) [reply sdk/nodejs/introspector/test/analysis.spec.ts]: Resolved in [27477bc](https://github.com/dagger/dagger/pull/6159/commits/27477bc051006373f25a0ad00364ef1e6bc95bea)

File: sdk/nodejs/introspector/decorators/decorators.ts
Comment (TomChv) [reply sdk/nodejs/introspector/decorators/decorators.ts:1]: Resolved in [27477bc](https://github.com/dagger/dagger/pull/6159/commits/27477bc051006373f25a0ad00364ef1e6bc95bea) with a registry logic

File: sdk/nodejs/introspector/analysis/analysis.ts (hunk 1 above)
Comment (TomChv) [reply sdk/nodejs/introspector/analysis/analysis.ts]: Resolved in [27477bc](https://github.com/dagger/dagger/pull/6159/commits/27477bc051006373f25a0ad00364ef1e6bc95bea)

File: sdk/nodejs/introspector/decorators/decorators.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (TomChv) [reply sdk/nodejs/introspector/decorators/decorators.ts]: Resolved in [27477bc](https://github.com/dagger/dagger/pull/6159/commits/27477bc051006373f25a0ad00364ef1e6bc95bea)

**Comment by TomChv** [comment]
@helderco After adding the registry logic in [27477bc](https://github.com/dagger/dagger/pull/6159/commits/27477bc051006373f25a0ad00364ef1e6bc95bea), I added the loading and invoke logic that will be used by the entrypoint.

The only missing part is the Runtime integration so I can call dagger module specific function to fill `invoke` parameter or load module if necessary.

We have never been that close to the integration of Node to Zenith!

I'll not add more feature to this PR, I want to add the runtime in a follow up PR 🚀

File: sdk/nodejs/introspector/test/invoke.spec.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+          inputBase.state,
+          inputBase.inputs
+        )
+
+        const inputInstall = {
+          objectName: "Alpine",
+          methodName: "install",
+          // Would be fetched from dagger and loaded from the container ID
+          state: inputBaseResult,
+          inputs: [["jq"]],
```
Comment (helderco) [inline sdk/nodejs/introspector/test/invoke.spec.ts:110]: Why are the inputs a list of values? Shouldn't it be an object?
```ts
inputs: {pkgs: ["jq"]},
```
Otherwise you'll need to sort the order from the API to match, not to mention missing or unset values.

File: sdk/nodejs/introspector/test/registry.spec.ts
Comment (helderco) [inline sdk/nodejs/introspector/test/registry.spec.ts:1]: You can also test for constructors (empty function name: `""`):
```ts
await registry.getResult("HelloWorld", "", {}, ["Hey"])
```
The above should call 
```ts
HelloWorld("Hey")
```

File: sdk/nodejs/introspector/test/testdata/state/state.ts (hunk 4 above)
Comment (helderco) [reply sdk/nodejs/introspector/test/testdata/state/state.ts:30]: Not actually resolved but can be in follow-up though. 🙂

**Comment by helderco** [comment]
I like the plain data structure you have btw, but with a small difference. I'd structure it to match core so that there's a 1:1 relationship between the data and the calls for registering. I've thought about doing that because of tests. It makes it much easier to test the registration phase if you can do a simple equals check in unit tests. Otherwise you have to mock the API calls to see if the right things are called.

File: sdk/nodejs/introspector/test/invoke.spec.ts (hunk 8 above)
Comment (TomChv) [reply sdk/nodejs/introspector/test/invoke.spec.ts:110]: Unset or missing value should return undefinied right. and the order is made by the introspector, it gives me the right order of argument

File: sdk/nodejs/introspector/registry/registry.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    if (!resolver.methods.find((m) => m === method)) {
+      throw new UnknownDaggerError(
+        `${method} is not registered in the resolver ${object}`,
+        {}
+      )
+    }
+
+    // Instantiate the class
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    let r = new resolver.class_() as any
```
[the remark this answers is not in the document]
Comment (TomChv) [reply sdk/nodejs/introspector/registry/registry.ts:135]: I expect like in Go to only be composed of optional value, this way I can reinject the previous state if a user call a constructor then a function.
